gitpod-io/gitpod
# [image-builder] Rebuild when upstream has changed

With this commit, the workspace image will be rebuilt when the upstream image (FROM in the .gitpod.Dockerfile) has changed. For this, the server extracts the FROM part of the Dockerfile and the image builder resolves the FROM image and adds the “digest” form of the image to the manifest hash that is used as image tag.

**How to test**

You can test this approach with an image in a registry that is under your control. E.g. push `gitpod/workspace-full` to `your-account/workspace-full` to Docker Hub. Add a repo with the following `.gitpod.Dockerfile`:
```
FROM your-account/workspace-full
RUN date > $HOME/date.txt
```

Open your repo in Gitpod. The workspace image will be built.

Change the image `your-account/workspace-full` on Docker Hub (e.g. push another/older tag of `workspace-full` as `latest`). Open another workspace with your repo without touching the repo content. The workspace image should be rebuilt.

**Impact of this change**

With this change, every workspace image based on a custom `.gitpod.Dockerfile` will be rebuilt since the way the hash tag is calculated has changed (by adding the upstream ref). This is somehow intentional because we want to rebuild with the latest upstream image and don't know if the workspace image is already on the latest upstream (after this change/rebuild we will know this for every workspace image).

We hope that this change will improve image pull times because this would reduce the number of different upstream images (e.g. the reference to `workspace-full` would always pull the latest `workspace-full` layers). However, that will also lead to more frequent image builds (every time an upstream image changes). Repo owners can still prevent this by using a specific image tag in FROM, e.g.:
```
docker.io/gitpod/workspace-full:latest@sha256:42181fe6917aedc92d61d9fbcfbfe28feb5c4d7b5db27cf6817e7a7c2b9df573
```
instead of just `gitpod/workspace-full`.

In fact, this behavior is more as I would expect it. When I don't pin to a specific version, I always get the latest upstream image. If I don't want this, I need to specify a specific image tag.

That also means Gitpod users don't need an artificial rebuild change (e.g. rebuild counter env var etc.) in their Dockerfile to be able to rebuild their image when the upstream image has changed.

A sample repo that is very active on Gitpod but uses a 3-month-old `workspace-full` reference is: https://github.com/tendermint/starport/blob/develop/.gitpod.Dockerfile – Workspace starts for this repo take usually ~60 seconds instead of ~10-20 seconds for repos with latest `workspace-full` image.
